toolchain/check/handle_class.cpp

```diff
@@ -642,46 +637,84 @@ static auto CheckCompleteAdapterClassType(Context& context,
       {.type_id = context.GetBuiltinType(SemIR::BuiltinInstKind::WitnessType),
        .object_repr_id = adapted_type_id});
 }
+static auto AddStructTypeFields(
+    Context& context,
+    llvm::SmallVector<SemIR::StructTypeField>& struct_type_fields) {
+  for (auto field_decl_id : context.field_decls_stack().PeekArray()) {
+    auto field_decl = context.insts().GetAs<SemIR::FieldDecl>(field_decl_id);
+    field_decl.index =
+        SemIR::ElementIndex{static_cast<int>(struct_type_fields.size())};
+    context.sem_ir().insts().Set(field_decl_id, field_decl);
+    if (field_decl.type_id == SemIR::TypeId::Error) {
+      struct_type_fields.push_back(
+          {.name_id = field_decl.name_id, .type_id = SemIR::TypeId::Error});
+      continue;
+    }
+    auto unbound_element_type =
+        context.sem_ir().types().GetAs<SemIR::UnboundElementType>(
+            field_decl.type_id);
+    struct_type_fields.push_back(
+        {.name_id = field_decl.name_id,
+         .type_id = unbound_element_type.element_type_id});
+  }
+  auto fields_id =
+      context.struct_type_fields().AddCanonical(struct_type_fields);
+  context.field_decls_stack().PopArray();
+  return fields_id;
+}
 
 // Checks that the specified finished class definition is valid and builds and
 // returns a corresponding complete type witness instruction.
 static auto CheckCompleteClassType(Context& context, Parse::NodeId node_id,
                                    SemIR::ClassId class_id) -> SemIR::InstId {
   auto& class_info = context.classes().Get(class_id);
   if (class_info.adapt_id.is_valid()) {
-    auto fields_id = context.struct_type_fields().AddCanonical(
-        context.struct_type_fields_stack().PeekArray());
-    context.struct_type_fields_stack().PopArray();
+    llvm::SmallVector<SemIR::StructTypeField> struct_type_fields;
+    struct_type_fields.reserve(context.field_decls_stack().PeekArray().size());
 
-    return CheckCompleteAdapterClassType(context, node_id, class_id, fields_id);
+    return CheckCompleteAdapterClassType(
+        context, node_id, class_id,
+        AddStructTypeFields(context, struct_type_fields));
```

Comment:
We're going to a bit of effort to build a collection of fields for an adapter here (and adapters can't have fields). I wonder if we could sink this field handling code down into `CheckCompleteAdapterClassType` and have it just look at and diagnose the first element in the top of `field_decls_stack` if it's non-empty (and in any case pop it).

Reply:
Oh, sure - b53b6b1441f0b8b12efc9430df0d021047bae633 - it has a minor observable sem-ir output change, which is that the field decls for the adapter have invalid indexes, since they never get updated/set by `AddstructTypeFields` - since the fields are invalid anyway, that seems fine?

Reply:
Yeah, I think that's fine. We might need to be slightly careful downstream to check for this case before indexing into the representation, but maybe that's necessary anyway to check for use while the class is incomplete.